guides/internals.rst

```diff
@@ -0,0 +1,342 @@
+===================================
+How to Work on Hypothesis Internals
+===================================
+
+This is a guide to how to work on Hypothesis internals,
+with a particular focus on helping people who are new to it.
+Right now it is very rudimentary and is intended primarily for people who are
+looking to get started writing shrink passes as part of our `current outreach
+program to get more people doing that <https://github.com/HypothesisWorks/hypothesis-python/issues/1093>`_,
+but it will expand over time.
+
+------------------------
+Bird's Eye View Concepts
+------------------------
+
+The core engine of Hypothesis is called Conjecture.
+
+The "fundamental idea" of Conjecture is that you can represent an arbitrary
+randomized test case as a string of bytes, which are basically intended as the
+underlying entropy of some pseudo-random number generator (PRNG).
+Whenever you want to do something "random" you read the next bytes and
+do what they tell you to do. By manipulating these bytes, we can achieve
+more interesting effects than pure randomness would allow us to do, while
+retaining the power and ease of use of random testing.
+
+The idea of shrinking in particular is that once we have this representation,
+we can shrink arbitrary test cases based on it. We try to produce a string that
+is *shortlex minimal*. What this means is that it has the shortest possible
+length and among those strings of minimal length is lexicographically (i.e. the
+normal order on strings - find the first byte at which they differ and use that
+to decide) smallest.
+
+Ideally we could think of the shrinker as a generic function that takes a
+string satisfying some predicate and returns the shortlex minimal string that
+also satisfies it.
+This is wrong on several levels: The first is that we only succeed in approximating
```

Comment:
"The implementation departs from this ideal in two ways:" ...

(IMO it's not *wrong* per se, just, well, idealised)